test/LiquidityHelper.sol

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+
+import {TestHelper, IERC20, Balances, Call, MockToken, Well, console} from "test/TestHelper.sol";
+import {MockFunctionBad} from "mocks/functions/MockFunctionBad.sol";
+import {IWellFunction} from "src/interfaces/IWellFunction.sol";
+
+/**
+ * @dev Stores the expected change in balance for User & Well throughout adding liquidity.
+ */
+struct AddLiquidityAction {
+    uint[] amounts;
+    uint[] fees;
+    uint lpAmountOut;
+    address recipient;
+}
+
+/**
+ * @dev Stores the expected change in balance for User & Well throughout remove liquidity.
+ */
+struct RemoveLiquidityAction {
+    uint[] amounts;
+    uint[] fees;
+    uint lpAmountIn;
+    address recipient;
+}
+
+/**
+ * @dev Holds a snapshot of User & Well balances. Used to calculate the change
+ * in balanace across some action in the Well.
+ */
+struct LiquiditySnapshot {
+    Balances user;
+    Balances well;
+    uint[] reserves;
+}
+
+/**
+ * @dev Provides common assertions when testing adding and removing liquidity.
+ *
+ * NOTE: Uses globals inherited from TestHelper.
+ */
+contract LiquidityHelper is TestHelper {
+    event AddLiquidity(uint[] amounts, uint lpAmountOut, address recipient);
+    event RemoveLiquidity(uint lpAmountIn, uint[] tokenAmountsOut, address recipient);
+
+    function beforeAddLiquidity(
+        uint[] memory amounts,
+        uint lpAmountOut,
+        address recipient
+    ) internal returns (LiquiditySnapshot memory, AddLiquidityAction memory) {
+        AddLiquidityAction memory action;
+
+        action.amounts = amounts;
+        action.lpAmountOut = lpAmountOut;
+        action.recipient = recipient;
+
+        return beforeAddLiquidity(action);
+    }
+
+    function beforeAddLiquidity(AddLiquidityAction memory action)
+        internal
+        returns (LiquiditySnapshot memory, AddLiquidityAction memory)
+    {
+        LiquiditySnapshot memory beforeSnapshot = _newSnapshot();
+
+        vm.expectEmit(true, true, true, false);
```

Comment:
Ideally we could change to this:

```suggestion
        vm.expectEmit(true, true, true, true, address(well));
```

However, we'll need to tweak the next line so that `action.amounts` can take into account fees.

I'd propose the following:
1. Make `LiquidityAction.fees` the amount of fee incurred, not the amount input after fees.
2. Add a helper method `_amountsWithFees(action)` which just performs `action.amounts[i] - action.fees[i]`.
3. Use `_amountsWithFees(action)` as a replacement for action.amounts. Then fees should be accounted for everywhere if they're applied by a test.

Reply:
Noted. I think for this we could just as equally, check if there are any fees like we do in the `afterAddLiquidity`

latest change includes this + keeps all the tests as they currently are, lmk what you think
```
        uint[] memory amountToTransfer = new uint[](tokens.length);
        for (uint i = 0; i < tokens.length; i++) {
            amountToTransfer[i] = action.fees[i] > 0 ? action.fees[i] : action.amounts[i];
        }

        vm.expectEmit(true, true, true, true, address(well));
        emit AddLiquidity(amountToTransfer, action.lpAmountOut, action.recipient);
```